File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/flutter/AdsEventForwarder.kt
package com.theoplayer.flutter

import com.theoplayer.android.api.ads.Ads
import com.theoplayer.android.api.event.EventListener
import com.theoplayer.android.api.event.ads.AdBeginEvent
import com.theoplayer.android.api.event.ads.AdBreakBeginEvent
import com.theoplayer.android.api.event.ads.AdBreakChangeEvent
import com.theoplayer.android.api.event.ads.AdBreakEndEvent
import com.theoplayer.android.api.event.ads.AdClickedEvent
import com.theoplayer.android.api.event.ads.AdEndEvent
import com.theoplayer.android.api.event.ads.AdErrorEvent
import com.theoplayer.android.api.event.ads.AdFirstQuartileEvent
import com.theoplayer.android.api.event.ads.AdImpressionEvent
import com.theoplayer.android.api.event.ads.AdLoadedEvent
import com.theoplayer.android.api.event.ads.AdMidpointEvent
import com.theoplayer.android.api.event.ads.AdSkipEvent
import com.theoplayer.android.api.event.ads.AdTappedEvent
import com.theoplayer.android.api.event.ads.AdThirdQuartileEvent
import com.theoplayer.android.api.event.ads.AddAdBreakEvent
import com.theoplayer.android.api.event.ads.AddAdEvent
import com.theoplayer.android.api.event.ads.AdsEventTypes
import com.theoplayer.android.api.event.ads.RemoveAdBreakEvent
import com.theoplayer.flutter.pigeon.THEOplayerFlutterAdsAPI
import com.theoplayer.flutter.transformers.AdsTransformer


class AdsEventForwarder(
    private val ads: Ads,
    private val flutterAPI: THEOplayerFlutterAdsAPI
) {
    private val emptyCallback: (Result<Unit>) -> Unit = {}

    private val adBeginEvent = EventListener<AdBeginEvent> {
        flutterAPI.onAdBegin(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! }, // if the param is not null, the return is never null
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! }, // if the param is not null, the return is never null
            callback = emptyCallback,
        )
    }

    private val adEndEvent = EventListener<AdEndEvent> {
        flutterAPI.onAdEnd(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }


    private val adErrorEvent = EventListener<AdErrorEvent> {
        flutterAPI.onAdError(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adMidpointEvent = EventListener<AdMidpointEvent> {
        flutterAPI.onAdMidpoint(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adAddedEvent = EventListener<AddAdEvent> {
        flutterAPI.onAddAd(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }


    private val adSkipEvent = EventListener<AdSkipEvent> {
        flutterAPI.onAdSkip(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adLoadedEvent = EventListener<AdLoadedEvent> {
        flutterAPI.onAdLoaded(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adTappedEvent = EventListener<AdTappedEvent> {
        flutterAPI.onAdTapped(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adClickedEvent = EventListener<AdClickedEvent> {
        flutterAPI.onAdClicked(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adFirstQuartileEvent = EventListener<AdFirstQuartileEvent> {
        flutterAPI.onAdFirstQuartile(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adThirdQuartileEvent = EventListener<AdThirdQuartileEvent> {
        flutterAPI.onAdThirdQuartile(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adImpressionEvent = EventListener<AdImpressionEvent> {
        flutterAPI.onAdImpression(
            adArg = AdsTransformer.toFlutterAd(it.ad),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }


    private val adBreakEndEvent = EventListener<AdBreakEndEvent> {
        flutterAPI.onAdBreakEnd(
            AdsTransformer.toFlutterAdBreak(it.adBreak),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adBreakBeginEvent = EventListener<AdBreakBeginEvent> {
        flutterAPI.onAdBreakBegin(
            AdsTransformer.toFlutterAdBreak(it.adBreak),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! }, // if the param is not null, the return is never null
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! }, // if the param is not null, the return is never null
            callback = emptyCallback,
        )
    }

    private val addAdBreakEvent = EventListener<AddAdBreakEvent> {
        flutterAPI.onAddAdBreak(
            AdsTransformer.toFlutterAdBreak(it.adBreak),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! }, // if the param is not null, the return is never null
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! }, // if the param is not null, the return is never null
            callback = emptyCallback,
        )
    }

    private val adBreakChangeEvent = EventListener<AdBreakChangeEvent> {
        flutterAPI.onAdBreakChange(
            AdsTransformer.toFlutterAdBreak(it.adBreak),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }

    private val adBreakRemoveEvent = EventListener<RemoveAdBreakEvent> {
        flutterAPI.onRemoveAdBreak(
            AdsTransformer.toFlutterAdBreak(it.adBreak),
            currentAdsArg = ads.currentAds.map { AdsTransformer.toFlutterAd(it)!! },
            currentAdBreakArg = ads.currentAdBreak?.let { AdsTransformer.toFlutterAdBreak(it) },
            scheduledAdsArg = ads.scheduledAds.map { AdsTransformer.toFlutterAd(it)!! },
            callback = emptyCallback,
        )
    }


    fun attachListeners() {
        ads.addEventListener(AdsEventTypes.ADD_AD, adAddedEvent)
        ads.addEventListener(AdsEventTypes.AD_BEGIN, adBeginEvent)
        ads.addEventListener(AdsEventTypes.AD_END, adEndEvent)
        ads.addEventListener(AdsEventTypes.AD_ERROR, adErrorEvent)
        ads.addEventListener(AdsEventTypes.AD_MIDPOINT, adMidpointEvent)
        ads.addEventListener(AdsEventTypes.AD_SKIP, adSkipEvent)
        ads.addEventListener(AdsEventTypes.AD_LOADED, adLoadedEvent)
        ads.addEventListener(AdsEventTypes.AD_TAPPED, adTappedEvent)
        ads.addEventListener(AdsEventTypes.AD_CLICKED, adClickedEvent)
        ads.addEventListener(AdsEventTypes.AD_FIRST_QUARTILE, adFirstQuartileEvent)
        ads.addEventListener(AdsEventTypes.AD_THIRD_QUARTILE, adThirdQuartileEvent)
        ads.addEventListener(AdsEventTypes.AD_IMPRESSION, adImpressionEvent)

        ads.addEventListener(AdsEventTypes.AD_BREAK_END, adBreakEndEvent)
        ads.addEventListener(AdsEventTypes.AD_BREAK_BEGIN, adBreakBeginEvent)
        ads.addEventListener(AdsEventTypes.ADD_AD_BREAK, addAdBreakEvent)
        ads.addEventListener(AdsEventTypes.AD_BREAK_CHANGE, adBreakChangeEvent)
        ads.addEventListener(AdsEventTypes.REMOVE_AD_BREAK, adBreakRemoveEvent)


    }

    fun removeListeners() {
        ads.removeEventListener(AdsEventTypes.ADD_AD, adAddedEvent)
        ads.removeEventListener(AdsEventTypes.AD_BEGIN, adBeginEvent)
        ads.removeEventListener(AdsEventTypes.AD_END, adEndEvent)
        ads.removeEventListener(AdsEventTypes.AD_ERROR, adErrorEvent)
        ads.removeEventListener(AdsEventTypes.AD_MIDPOINT, adMidpointEvent)
        ads.removeEventListener(AdsEventTypes.AD_SKIP, adSkipEvent)
        ads.removeEventListener(AdsEventTypes.AD_LOADED, adLoadedEvent)
        ads.removeEventListener(AdsEventTypes.AD_TAPPED, adTappedEvent)
        ads.removeEventListener(AdsEventTypes.AD_CLICKED, adClickedEvent)
        ads.removeEventListener(AdsEventTypes.AD_FIRST_QUARTILE, adFirstQuartileEvent)
        ads.removeEventListener(AdsEventTypes.AD_THIRD_QUARTILE, adThirdQuartileEvent)
        ads.removeEventListener(AdsEventTypes.AD_IMPRESSION, adImpressionEvent)

        ads.removeEventListener(AdsEventTypes.AD_BREAK_END, adBreakEndEvent)
        ads.removeEventListener(AdsEventTypes.AD_BREAK_BEGIN, adBreakBeginEvent)
        ads.removeEventListener(AdsEventTypes.ADD_AD_BREAK, addAdBreakEvent)
        ads.removeEventListener(AdsEventTypes.AD_BREAK_CHANGE, adBreakChangeEvent)
        ads.removeEventListener(AdsEventTypes.REMOVE_AD_BREAK, adBreakRemoveEvent)

    }


}

File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_ios/ios/Classes/AdsEventForwarder.swift
//
//  AdsEvenrForwarder.swift
//  Pods
//
//  Created by Dallos, Daniel on 08/01/2025.
//

import Foundation
import Flutter
import THEOplayerSDK

class AdsEventForwarder {
 
    private let ads: Ads
    private let flutterAPI: THEOplayerFlutterAdsAPI
    
    private var adBeginEventListener: EventListener?
    private var adEndEventListener: EventListener?
    private var adErrorEventListener: EventListener?
    private var adFirstQuartileEventListener: EventListener?
    private var adMidpointEventListener: EventListener?
    private var adThirdQuartileEventListener: EventListener?
    private var adLoadedEventListener: EventListener?
    private var adImpressionEventListener: EventListener?
    private var adSkipEventListener: EventListener?
    private var addAdEventListener: EventListener?

    private var adBreakBeginEventListener: EventListener?
    private var adBreakEndEventListener: EventListener?

    private var adTappedEventListener: EventListener?
    private var adClickedEventListener: EventListener?

    private var addAdBreakEventListener: EventListener?
    private var removeAdBreakEventListener: EventListener?
    private var adBreakChangeEventListeners: EventListener?
    
    //TODO: for SGAI
    //private var updateAdEventListener: EventListener?
    //private var updateAdBreakEventListener: EventListener?

    
    
    init(ads: Ads, flutterAPI: THEOplayerFlutterAdsAPI) {
        self.ads = ads
        self.flutterAPI = flutterAPI
    }
    
    func attachListeners() {
        
        let emptyCompletion: (Result<Void, FlutterError>) -> Void = {result in }

        adBeginEventListener = ads.addEventListener(type: AdsEventTypes.AD_BEGIN) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdBegin(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        adEndEventListener = ads.addEventListener(type: AdsEventTypes.AD_END) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdEnd(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        adErrorEventListener = ads.addEventListener(type: AdsEventTypes.AD_ERROR) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdError(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

        adFirstQuartileEventListener = ads.addEventListener(type: AdsEventTypes.AD_FIRST_QUARTILE) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdFirstQuartile(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        adMidpointEventListener = ads.addEventListener(type: AdsEventTypes.AD_MIDPOINT) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdMidpoint(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

        adThirdQuartileEventListener = ads.addEventListener(type: AdsEventTypes.AD_THIRD_QUARTILE) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdThirdQuartile(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

        adLoadedEventListener = ads.addEventListener(type: AdsEventTypes.AD_LOADED) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdLoaded(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        adImpressionEventListener = ads.addEventListener(type: AdsEventTypes.AD_IMPRESSION) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdImpression(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

        adSkipEventListener = ads.addEventListener(type: AdsEventTypes.AD_SKIP) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdSkip(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        adTappedEventListener = ads.addEventListener(type: AdsEventTypes.AD_TAPPED) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdTapped(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

        adClickedEventListener = ads.addEventListener(type: AdsEventTypes.AD_CLICKED) { [weak self] event in
            guard let welf = self else { return }

            welf.flutterAPI.onAdClicked(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        addAdEventListener = ads.addEventListener(type: AdsEventTypes.ADD_AD) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAddAd(
                ad: AdsTransformer.toFlutterAd(ad: event.ad),
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        adBreakBeginEventListener = ads.addEventListener(type: AdsEventTypes.AD_BREAK_BEGIN) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdBreakBegin(
                adbreak: AdsTransformer.toFlutterAdBreak(adBreak: event.ad)!,
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

        adBreakEndEventListener = ads.addEventListener(type: AdsEventTypes.AD_BREAK_END) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdBreakEnd(
                adbreak: AdsTransformer.toFlutterAdBreak(adBreak: event.ad)!,
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        addAdBreakEventListener = ads.addEventListener(type: AdsEventTypes.ADD_AD_BREAK) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAddAdBreak(
                adbreak: AdsTransformer.toFlutterAdBreak(adBreak: event.ad)!,
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }
        
        removeAdBreakEventListener = ads.addEventListener(type: AdsEventTypes.REMOVE_AD_BREAK) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onRemoveAdBreak(
                adbreak: AdsTransformer.toFlutterAdBreak(adBreak: event.ad)!,
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

        adBreakChangeEventListeners = ads.addEventListener(type: AdsEventTypes.AD_BREAK_CHANGE) { [weak self] event in
            guard let welf = self else { return }
            
            welf.flutterAPI.onAdBreakChange(
                adbreak: AdsTransformer.toFlutterAdBreak(adBreak: event.ad)!,
                currentAds: welf.ads.currentAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                currentAdBreak: AdsTransformer.toFlutterAdBreak(adBreak: welf.ads.currentAdBreak),
                scheduledAds: welf.ads.scheduledAds.map({ ad in
                    AdsTransformer.toFlutterAd(ad: ad)! // parameter can not be null, so the return will be never null
                }),
                completion: emptyCompletion)
        }

    }
        
    func detachListeners(){
        ads.removeEventListener(type: AdsEventTypes.AD_BEGIN, listener: adBeginEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_END, listener: adEndEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_ERROR, listener: adErrorEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_FIRST_QUARTILE, listener: adFirstQuartileEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_MIDPOINT, listener: adMidpointEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_THIRD_QUARTILE, listener: adThirdQuartileEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_LOADED, listener: adLoadedEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_IMPRESSION, listener: adImpressionEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_SKIP, listener: adSkipEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_BREAK_BEGIN, listener: adBreakBeginEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_BREAK_END, listener: adBreakEndEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_TAPPED, listener: adTappedEventListener!)
        ads.removeEventListener(type: AdsEventTypes.AD_CLICKED, listener: adClickedEventListener!)
        ads.removeEventListener(type: AdsEventTypes.ADD_AD_BREAK, listener: addAdBreakEventListener!)
        ads.removeEventListener(type: AdsEventTypes.REMOVE_AD_BREAK, listener: removeAdBreakEventListener!)
    }
    
}
